app/src/main/java/com/kamwithk/ankiconnectandroid/ankidroid_api/IntegratedAPI.java

```diff
@@ -67,78 +68,176 @@ public void addSampleCard() {
 
-        if (sameModelName) {
-            Map<String, Long> modelNameToId = modelAPI.modelNamesAndIds(0);
-            Long modelId = modelNameToId.get(modelName);
-            if (modelId == null) { // i.e. not found
-                // all false! (cannot add the note if there's no valid model to add it with)
-                ArrayList<Boolean> allFalse = new ArrayList<>();
-                for (int i = 0; i < notesToTest.size(); i++) {
-                    allFalse.add(false);
-                }
-                return allFalse;
+        //If duplicates are allowed, just need to see if they are valid notes (checksum != 0)
+        if (noteOptions.isAllowDuplicate()) {
+            for (long checksum: checksums) {
+                canAddNote.add(checksum != 0);
             }
+            return canAddNote;
+        }
 
-            // Otherwise, we finally can use the internal API call
-            List<String> keys = new ArrayList<>();
-            for (Parser.NoteFront noteFront : notesToTest) {
-                keys.add(noteFront.getFieldValue());
-            }
-            SparseArray<List<NoteInfo>> duplicateNotes =
-                    api.findDuplicateNotes(modelId, keys);
+        //Grabbing the note options and model for the first note and assuming the rest are the same.
+        //This is true for yomitan but might not be for other applications.
+        String modelName = notesToTest.get(0).getModelName();
 
-            ArrayList<Boolean> noteDoesNotExist = new ArrayList<>();
+        //If duplicate scope is "deck" or "deck root", we need to get extra information to figure out if did matches.
+        //If duplicate scope is "deck root" we need to include children, noteOptions.getDeckName() will not be null
+        String deckName = noteOptions.getDeckName();
+        HashSet<Long> deckIds = new HashSet<>();
+        Map<String, Long> deckNamesToIds = deckAPI.deckNamesAndIds();
+        if(deckName == null) {
+            //Deck, not root
+            deckName = notesToTest.get(0).getDeckName();
+            deckIds.add(deckNamesToIds.get(deckName));
+        }
+        else {
+            for (String name: deckNamesToIds.keySet()) {
+                if (name.contains(deckName)) {
+                    deckIds.add(deckNamesToIds.get(name));
+                }
+            }
+        }
+        Map<String, Long> modelNameToId = modelAPI.modelNamesAndIds(0);
+        Long modelId = modelNameToId.get(modelName);
+
+        String selectionQuery = "";
+        if (!noteOptions.isCheckAllModels()) {
+            selectionQuery = String.format(
+                    Locale.US,
+                    "%s=%d and ",
+                    FlashCardsContract.Note.MID,
+                    modelId
+            );
+        }
+        selectionQuery = selectionQuery + String.format(
+                Locale.US,
+                "%s in (%s)",
+                FlashCardsContract.Note.CSUM,
+                TextUtils.join(",", checksums)
+        );
```

Comment:
This is what AnkiDroid does internally for its duplicate note check, since we're checking for the IDs of the model and notes it shouldn't have an effect.